File: LeetCode/1143/solution.recursive-dp.java
// 1143. Longest Common Subsequence
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int[][]dp = new int[text1.length()+1][text2.length()+1];
        for (int i = 0; i < dp.length; i++) {
            Arrays.fill(dp[i], -1);
        }

        return solve(0, 0, text1, text2, dp);
    }

    private int solve(int i, int j, String str1, String str2, int[][] dp){
        if(i==str1.length() || j==str2.length()) return 0;
        if(dp[i][j]!=-1) return dp[i][j];

        int ans = 0;
        if(str1.charAt(i)==str2.charAt(j)) {
            ans = 1 + solve(i+1, j+1, str1, str2, dp);
        }
        else {
            int ret1 = solve(i+1, j, str1, str2, dp);
            int ret2 = solve(i, j+1, str1, str2, dp);
            ans = Math.max(ret1, ret2);
        }

        return dp[i][j] = ans;
    }
}

File: LeetCode/1143/solution.recursive-dp.cpp
// 1143. Longest Common Subsequence
class Solution {
public:
    string str1, str2;
    int len1, len2;
    vector<vector<int>> dp;

    int longestCommonSubsequence(string text1, string text2) {
        str1 = text1;
        str2 = text2;
        len1 = text1.size();
        len2 = text2.size();

        dp.assign(len1, vector(len2, -1));

        return solve(0, 0);
    }

    int solve(int i, int j){
        if(i==len1 || j==len2) return 0;
        if(dp[i][j]!=-1) return dp[i][j];

        int ans=0;
        if(str1[i]==str2[j]) 
            ans = 1 + solve(i+1, j+1);
        else 
            ans = max(solve(i+1, j), solve(i, j+1));

        return dp[i][j] = ans;
    }
};